modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/transaction/TransactionService.scala

```diff
@@ -3,37 +3,65 @@ package org.tessellation.dag.l1.domain.transaction
 import cats.data.NonEmptyList
 import cats.data.Validated.{Invalid, Valid}
 import cats.effect.Async
-import cats.syntax.applicative._
-import cats.syntax.either._
-import cats.syntax.flatMap._
-import cats.syntax.functor._
+import cats.syntax.all._
 
-import org.tessellation.node.shared.domain.transaction.ContextualTransactionValidator
-import org.tessellation.node.shared.domain.transaction.ContextualTransactionValidator.ContextualTransactionValidationError
+import org.tessellation.dag.l1.domain.transaction.ContextualTransactionValidator.NonContextualValidationError
+import org.tessellation.ext.cats.syntax.validated.validatedSyntax
+import org.tessellation.node.shared.domain.collateral.LatestBalances
+import org.tessellation.node.shared.domain.snapshot.storage.LastSnapshotStorage
+import org.tessellation.node.shared.domain.transaction.TransactionValidator
+import org.tessellation.schema.SnapshotOrdinal
+import org.tessellation.schema.balance.Balance
+import org.tessellation.schema.snapshot.{Snapshot, SnapshotInfo, StateProof}
 import org.tessellation.schema.transaction.Transaction
 import org.tessellation.security.Hashed
 import org.tessellation.security.hash.Hash
 
+import fs2.Stream
+
+import ContextualTransactionValidator.ContextualTransactionValidationError
+
 trait TransactionService[F[_]] {
   def offer(transaction: Hashed[Transaction]): F[Either[NonEmptyList[ContextualTransactionValidationError], Hash]]
 }
 
 object TransactionService {
 
-  def make[F[_]: Async](
+  def make[
+    F[_]: Async,
+    P <: StateProof,
+    S <: Snapshot,
+    SI <: SnapshotInfo[P]
+  ](
     transactionStorage: TransactionStorage[F],
-    contextualTransactionValidator: ContextualTransactionValidator[F]
+    lastSnapshotStorage: LastSnapshotStorage[F, S, SI] with LatestBalances[F],
+    transactionValidator: TransactionValidator[F]
   ): TransactionService[F] = new TransactionService[F] {
 
     def offer(
       transaction: Hashed[Transaction]
     ): F[Either[NonEmptyList[ContextualTransactionValidationError], Hash]] =
-      contextualTransactionValidator.validate(transaction.signed).flatMap {
-        case Valid(_) =>
-          transactionStorage
-            .put(transaction)
-            .as(transaction.hash.asRight[NonEmptyList[ContextualTransactionValidationError]])
-        case Invalid(e) => e.toNonEmptyList.asLeft[Hash].pure[F]
+      lastSnapshotStorage.getCombinedStream.map {
+        case Some((s, si)) => (s.ordinal, si.balances.getOrElse(transaction.source, Balance.empty))
+        case None          => (SnapshotOrdinal.MinValue, Balance.empty)
+      }.changes.switchMap {
+        case (latestOrdinal, balance) =>
+          Stream.eval {
+            transactionValidator
+              .validate(transaction.signed)
```

Comment:
I guess this validate can happen only once given that it's a static validation. Not sure if worth refactoring

Reply:
Fair point, no need to rerun static validation if stream emits.